### PR TITLE
feat(image_projection_based_fusion): add objects filter by rois

### PR DESCRIPTION
**Pull request by yukke42**
## Description

summary

- change `passthrough_lower_bound_probability_threshold` can be used for each class
- add this pakcage after centerpoint or pointpainting to filter false positives, especially for pedestrian.

including <https://github.com/autowarefoundation/autoware.universe/pull/4057>

### behavior

- bounding box on image
  - blue ... projected DetectedObjects from centerpoint
  - red ... RoIs from YOLOX
- bounding box on bev 
  - blue ... output objects from this package
  - yellow ... false positives filtered out by this package

to show an image on screenshot, please change debug_mode=true on roi_detected_object_fusion.launch.xml

image

### thrust_distances

If the distance of obejct is greater than the threshold (50.0m for pedestrian), it's not discared by this package.

In the screenshot below, a pedestrian is filtered by this pacakge since YOLOX can not detect it even though  pointpainting can detect it correctly. After adding thrust distance, it cloud be output correctly from this package.

image

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### Summary by CodeRabbit

Release Notes:
- New Feature: Modified the signature of `generateDetectedObjectRoIs` and `fuseObjectsOnImage` functions to accept a `DetectedObjects` message instead of a vector of `DetectedObject` objects.
- New Feature: Added a new member variable `can_assign_matrix` of type `Eigen::MatrixXi`.
- New Feature: Modified member variables `passthrough_object_flags_`, `fused_object_flags_`, and `ignored_object_flags_` to be maps with `int64_t` keys instead of vectors.
- New Feature: Added a new argument `param_path` with a default value to the launch file.
- New Feature: Included several new parameters for rois, camera info, and images in the launch file.
- New Feature: The `roi_detected_object_fusion_node` now reads parameters from the launch file.
- New Feature: Set the `debug_mode` parameter to `true` by default.
- Chore: Added a new dependency on `object_recognition_utils` in the package manifest file.

> "With new signatures and maps so grand,
> Parameters and dependencies expand.
> Launch files read, debug mode set,
> A fusion of features, no regrets."

### CODE REVIEW

**Comment by yukke42** [comment]
I want to merge PR ASAP, so it's opened now. But I need to tune params.
This PR conflicts with main branch because it depends on #4057.

**Review by miursh — APPROVED** [review]
LGTM

File: perception/image_projection_based_fusion/src/roi_detected_object_fusion/node.cpp [hunk 1]
```diff
@@ -255,6 +320,10 @@ void RoiDetectedObjectFusionNode::publish(const DetectedObjects & output_msg)
 
   debug_publisher_->publish<DetectedObjects>("debug/fused_objects", debug_fused_objects_msg);
   debug_publisher_->publish<DetectedObjects>("debug/ignored_objects", debug_ignored_objects_msg);
+
+  passthrough_object_flags_map_.erase(timestamp_nsec);
+  fused_object_flags_map_.erase(timestamp_nsec);
+  fused_object_flags_map_.erase(timestamp_nsec);
```
Comment (miursh) [inline perception/image_projection_based_fusion/src/roi_detected_object_fusion/node.cpp:326]:
```suggestion
  ignored_object_flags_map_.erase(timestamp_nsec);
```

Comment (yukke42) [reply perception/image_projection_based_fusion/src/roi_detected_object_fusion/node.cpp]: thanks, applied at [d8394e9](https://github.com/autowarefoundation/autoware.universe/pull/4546/commits/d8394e9febf252dbd7fe106b6e8253c9939e5a8a)

**Comment by yukke42** [comment]
I have added thrust_distance and updated description.

File: perception/image_projection_based_fusion/include/image_projection_based_fusion/roi_detected_object_fusion/node.hpp [hunk 2]
```diff
@@ -56,10 +58,12 @@ class RoiDetectedObjectFusionNode : public FusionNode<DetectedObjects, DetectedO
 private:
   struct
   {
-    double passthrough_lower_bound_probability_threshold{};
+    std::vector<double> passthrough_lower_bound_probability_thresholds{};
+    std::vector<double> thrust_distances{};
```
[the remark this answers is not in the document]
Comment (yukke42) [reply perception/image_projection_based_fusion/include/image_projection_based_fusion/roi_detected_object_fusion/node.hpp]: Thank you. changed at [a246dd1](https://github.com/autowarefoundation/autoware.universe/pull/4546/commits/a246dd1a37d421f0033eb5e09840d41a2711a150).

File: perception/image_projection_based_fusion/config/roi_detected_object_fusion.param.yaml [hunk 3]
```diff
@@ -0,0 +1,18 @@
+/**:
+  ros__parameters:
+    # UNKNOWN, CAR, TRUCK, BUS, TRAILER, MOTORBIKE, BICYCLE, PEDESTRIAN
+    passthrough_lower_bound_probability_thresholds: [0.35, 0.35, 0.35, 0.35, 0.35, 0.35, 0.35, 0.50]
+    min_iou_threshold: 0.5
```
Comment (miursh) [inline perception/image_projection_based_fusion/config/roi_detected_object_fusion.param.yaml:5]: @yukke42 Need to add trust_distances in param.yaml ?

Comment (yukke42) [reply perception/image_projection_based_fusion/config/roi_detected_object_fusion.param.yaml:6]: sorry, I missed it. [4d68604](https://github.com/autowarefoundation/autoware.universe/pull/4546/commits/4d68604595875814bc4eda886f10b8c497a18766)

File: perception/image_projection_based_fusion/config/roi_detected_object_fusion.param.yaml [hunk 4]
```diff
@@ -0,0 +1,19 @@
+/**:
+  ros__parameters:
+    # UNKNOWN, CAR, TRUCK, BUS, TRAILER, MOTORBIKE, BICYCLE, PEDESTRIAN
+    passthrough_lower_bound_probability_thresholds: [0.35, 0.35, 0.35, 0.35, 0.35, 0.35, 0.35, 0.50]
+    thrust_distances: [50.0, 100.0, 100.0, 100.0, 100.0, 50.0, 50.0, 50.0]
```
Comment (miursh) [inline perception/image_projection_based_fusion/config/roi_detected_object_fusion.param.yaml:5]:
```suggestion
    trust_distances: [50.0, 100.0, 100.0, 100.0, 100.0, 50.0, 50.0, 50.0]
```

**Review by miursh — APPROVED** [review]
LGTM

**Review by yukkysaito — APPROVED** [review]
LGTM

**Pull request merged**